### ISSUE TITLE
Don’t show placeholder when loaded from cache

### ISSUE THREAD

**Issue by msw333** [renefloor/flutter_cached_network_image#164]
Maybe I am doing something wrong: it is perfect to show a placeholder, like a progress indicator, when an image is loaded from a web source.

But I don’t think it is a good idea to show a placeholder when images are loaded from cache, since they can appear directly.

Highly appreciate your comment on that.

**Comment by crush3r** [comment]
I agree.

If the image has already been downloaded once, even after restarting the application, it should be displayed immediately instead of the placeholder.

It is the last thing I need. It would be great!

**Comment by roidy** [comment]
I agree this needs fixing, it looks terrible in list based apps when a placeholder flashes up for a split second while the cached image is loaded. Placeholder should only be shown when getting an image across the network or if no image exists.

**Comment by cubissimo** [comment]
Maybe a simple new parameter allowing us to choose if we want or not this behavior

**Comment by renefloor** [comment]
Would you want to keep the fading animation in that case or not?

**Comment by hgl** [comment]
@renefloor I think if the image already exists in cache, it should display immediately just like a regular image, without any fading animation.

**Comment by KaiP-598** [comment]
Hi,
I have the same problem here, I am using this in the listview, and when I scroll up or down, although the images are cached, but there's still a split second before the image is shown.

**Comment by ejdrian313** [comment]
its very important to me to avoid placeholder on cached images. so is any of you found a solution for this?

**Comment by Pannkaaka** [comment]
Any updates on this? Kind of a deal breaker for me..

**Comment by royibernthal** [comment]
Joining this issue, a way to avoid that split second flash for an already cached image would be a great touch.

**Comment by SvehlaJan** [comment]
Hello. Any update on this? It is a deal breaker for me too.

**Comment by ejdrian313** [comment]
little workaround:
`import 'package:flutter/material.dart';

class ImageWidgetPlaceholder extends StatelessWidget {
  const ImageWidgetPlaceholder({Key key, this.image, this.placeholder})
      : super(key: key);

  final ImageProvider image;
  final Widget placeholder;

  @override
  Widget build(BuildContext context) {
    return Image(
      image: image,
      frameBuilder: (BuildContext context, Widget child, int frame,
          bool wasSynchronouslyLoaded) {
        if (wasSynchronouslyLoaded) {
          return child;
        } else {
          return AnimatedSwitcher(
            duration: Duration.zero,
            child: frame != null ? child : placeholder,
          );
        }
      },
    );
  }
}
`

with use in a container:
`Container(
      decoration: BoxDecoration(
        image: DecorationImage(
          image: CachedNetworkImageProvider(
            page['image'],
          ),
          fit: BoxFit.fill,
        ),
      ),
....`

or if you now size of the image avoid layout jump: 
`ImageWidgetPlaceholder(
                          image: CachedNetworkImageProvider(
                            multiselect['imageUrl'],
                          ),
                          placeholder: SizedBox(
                            width: MediaQuery.of(context).size.width,
                            height: 200,
                          ),
                        )`

**Comment by grantespo** [comment]
Amazing library, starred!

Unfortunately, this is a dealbreaker for me too

**Comment by cemozerr** [comment]
Any updates on this? This is a dealbreaker for me as well.

**Comment by cemozerr** [comment]
I dug deep into this issue and, correct me if I'm wrong @renefloor but, it seems like even if the image **information** is held in memory, the `FlutterCacheManager` holds the image bytes on disk. This seems to be why image can't be displayed instantaneously. 

Or generally, rather that the cache implementation the `CachedNetworkImage` uses does not have synchronous return methods for the image bytes that are stored either on disk or memory.

**Comment by royvangeel** [comment]
Anyone has an update on this? Got the same issue, don't want to see the progress indicator when loaded from cache

**Comment by zaid17** [comment]
Any updates on this?

**Comment by advancedngwarati** [comment]
Any progress on this? Will this be possible?

**Comment by creativecreatorormaybenot** [comment]
> I dug deep into this issue and, correct me if I'm wrong @renefloor but, it seems like even if the image information is held in memory, the FlutterCacheManager holds the image bytes on disk. This seems to be why the image can't be displayed instantaneously.

@cemozerr I don't know how you came up with this, but that is just wrong. The workaround @ejdrian313 showed above works and allows to display the images synchronously.

The issue is definitely a bug in the `CachedNetworkImage` widget. And of course the image bytes can be cached in memory. This is the reason the [`Image` widget has a `frameBuilder` parameter](https://api.flutter.dev/flutter/widgets/Image/frameBuilder.html) that allows you to respond to images loaded from memory via `wasSynchronouslyLoaded`. When caching images in Flutter's in-memory `ImageCache`, you can also scale down the resolution to reduce the byte size of the image in cache and the `ImageCache` has has max size and count properties.

**Comment by u-sour** [comment]
any solution?

**Comment by Erengun** [comment]
OctoImage's frameBuilder uses Flutter's wasSynchronouslyLoaded flag. Since disk-cached images still load asynchronously (file I/O + decode = async), this flag is always false, triggering the full placeholder + fade flow even for fast cache hits.
